pil-analyzer/src/evaluator.rs

```diff
@@ -837,20 +858,33 @@ impl<'a, 'b, T: FieldElement, S: SymbolLookup<'a, T>> Evaluator<'a, 'b, T, S> {
                                 })
                                 .unwrap();
 
-                            // let Expression::LambdaExpression(_, body) = &expr.as_ref() else {
```

Comment:
I think it would be nice to move all these things into the trait (i.e. just into the lookup function, then we don't need a separate function for trait impls). Maybe the only thing the evaulator should do is substitute the type args. Please check if it is better to change the signature of the lookup function to take a poly reference instead of just the name.